### PR TITLE
Extension Codestarts are now "really" meant to start coding with an extension

### LINKED ISSUE #15997: quarkus-maven-plugin create adds conflicting extensions: config-yaml adds quarkus-resteasy
## Describe the bug
When using the quarkus maven plugin, if you specify the "quarkus-config-yaml" extension, for some reason it automatically adds the "quarkus-resteasy" extension, which is probably because the project generator writes the ConfigResource example code and assumes it needs the extension. 

However, when specifying the extensions, if "quarkus-resteasy-reactive" is used, then the initial project creation ends up with conflicting extensions. 

### Expected behavior
I would expect that the generator would check if "quarkus-resteasy-reactive" is in the extension list, and if it does, then it would exclude the automatic inclusion of the "quarkus-resteasy" extension.

### Actual behavior
Both extensions are added and results in an error with conflicting reactive and blocking extensions.

## To Reproduce

mvn io.quarkus:quarkus-maven-plugin:1.12.2.Final:create \
    -DprojectGroupId=com.redhat.openshift \
    -DprojectArtifactId=project-organization-api \
    -DclassName="com.redhat.openshift.pmo.GroupResource" \
    -Dpath="/api/groups" \
    -Dextensions="config-yaml,resteasy-reactive,resteasy-reactive-jackson"

### Screenshots
```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.654 s <<< FAILURE! - in com.redhat.openshift.pmo.GroupResourceTest
[ERROR] com.redhat.openshift.pmo.GroupResourceTest.testHelloEndpoint  Time elapsed: 0.003 s  <<< ERROR!
java.lang.RuntimeException: java.lang.RuntimeException: java.lang.IllegalArgumentException: Multiple matching properties for name "security.jaxrs.deny-unannotated-endpoints" property was matched by both public boolean io.quarkus.resteasy.runtime.JaxRsSecurityConfig.denyJaxRs and public boolean io.quarkus.resteasy.reactive.common.runtime.JaxRsSecurityConfig.denyJaxRs. This is likely because you have an incompatible combination of extensions that both define the same properties (e.g. including both reactive and blocking database extensions)
Caused by: java.lang.RuntimeException: java.lang.IllegalArgumentException: Multiple matching properties for name "security.jaxrs.deny-unannotated-endpoints" property was matched by both public boolean io.quarkus.resteasy.runtime.JaxRsSecurityConfig.denyJaxRs and public boolean io.quarkus.resteasy.reactive.common.runtime.JaxRsSecurityConfig.denyJaxRs. This is likely because you have an incompatible combination of extensions that both define the same properties (e.g. including both reactive and blocking database extensions)
Caused by: java.lang.IllegalArgumentException: Multiple matching properties for name "security.jaxrs.deny-unannotated-endpoints" property was matched by both public boolean io.quarkus.resteasy.runtime.JaxRsSecurityConfig.denyJaxRs and public boolean io.quarkus.resteasy.reactive.common.runtime.JaxRsSecurityConfig.denyJaxRs. This is likely because you have an incompatible combination of extensions that both define the same properties (e.g. including both reactive and blocking database extensions)
```

### PR DESCRIPTION
**Pull request by ia3andy**
- Change from `--no-examples` to `--no-code` in Create command for Maven Plugin and CLI
- Change from `--examples` to `--example` in Create command for Maven Plugin and CLI (only one example can be selected at a time)
- some `example` are moved to `extension-codestarts` and rewritten to be starters (almost no logic): funqy-http, resteasy-qute, picocli, config-yaml and resteasy, resteasy-reactive, spring-web
- All `singleton-example` become `example`, only one can be activated for an app, they disable `extension-codestarts`
- `providesExample` becomes `providesCode` in `ExtensionProcessor`
- Remove `resteasy-jackson` and `logging-json` examples which are not really useful for now

THIS IS MEANT FOR 2.0 AS IT BREAKS COMPAT

Fixes #15997 (config-yaml is not using resteasy anymore, it just creates a `GreetingConfig`)

## Generated code:

resteasy, resteasy-reactive and spring-web are not changed.

### RESTEasy Qute

```java
package org.acme;

import ...

@Path("/some-page")
public class SomePage {

    @Inject
    Template page;

    @GET
    @Produces(MediaType.TEXT_HTML)
    public TemplateInstance get(@QueryParam("name") String name) {
        return page.data("name", name);
    }

}
```

### YAML Config

You have to bind it yourself somewhere to actually use it (command line, resteasy, resteasy-reactive)

```java
package org.acme;

import ...

@ConfigProperties(prefix = "greeting")
public interface GreetingConfig {

    @ConfigProperty(name = "message")
    String message();

}
```

### picocli

```java
package org.acme;

import ...

@Command(name = "greeting", mixinStandardHelpOptions = true)
public class GreetingCommand implements Runnable {

    @Parameters(paramLabel = "<name>", defaultValue = "picocli",
        description = "Your name.")
    String name;

    @Override
    public void run() {
        System.out.printf("Hello %s, go go commando!\n", name);
    }

}

```

### Funqy HTTP

```java
package org.acme;

import io.quarkus.funqy.Funq;

public class MyFunctions {

    @Funq
    public String greeting(GreetingInput input) {
        return String.format("Hello %s!", input != null ? input.name : "Funqy");
    }


    public static class GreetingInput {
        public String name;
    }

}

```

### CODE REVIEW

**Comment by ia3andy** [comment]
To be sure this is worth it, just create an app using some of those extensions: `funqy-http, resteasy-qute, picocli, config-yaml and resteasy, resteasy-reactive, spring-web` and you will see that they combine well are now really making you start quicker on your app.

**Comment by FroMage** [comment]
In the case of Qute, there's also the template file, right? In the case of YAML, there's also the YAML config file, right?

Otherwise, looks like the right move :)

**Comment by ia3andy** [comment]
> In the case of Qute, there's also the template file, right? In the case of YAML, there's also the YAML config file, right?
> 
> Otherwise, looks like the right move :)

Yes and yes! :)

```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>Hello {name ?: "Qute"}</title>
    <style>
        body {
            background-color: #000;
            color: #fff;
        }
        h1 {
            color: #4695EB;
            font-size: 6vw;
        }

        h1 b {
            color: #be9100;
        }

        p {
            font-size: 2vw;
        }
    </style>
</head>
<body>
<h1>Hello <b>{name ?: "Qute"}</b></h1>

<p>Create your web page using Quarkus RESTEasy & Qute</p>
</body>
</html>
```

and 

```yaml
greeting:
  message: "hello"

```

**Comment by ia3andy** [comment]
@unfromage
image

File: devtools/project-core-extension-codestarts/src/test/java/io/quarkus/devtools/codestarts/quarkus/FunqyHttpCodestartTest.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+            .codestarts("funqy-http")
+            .languages(JAVA)
+            .build();
+
+    @Test
+    void testContent() throws Throwable {
+        codestartTest.checkGeneratedSource("org.acme.MyFunctions");
+        codestartTest.checkGeneratedTestSource("org.acme.MyFunctionsTest");
+        codestartTest.checkGeneratedTestSource("org.acme.MyFunctionsIT");
+        //codestartTest.buildAllProjects();
```
Comment (glefloch) [inline devtools/project-core-extension-codestarts/src/test/java/io/quarkus/devtools/codestarts/quarkus/FunqyHttpCodestartTest.java:23]: This line is commented in multiple classes, should it be remove ?

File: independent-projects/tools/devtools-testing/src/main/java/io/quarkus/devtools/testing/RegistryClientTest.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package io.quarkus.devtools.testing;
+
+import org.junit.jupiter.api.extension.AfterAllCallback;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+public class RegistryClientTest implements BeforeAllCallback, AfterAllCallback {
```
Comment (aloubyansky) [inline independent-projects/tools/devtools-testing/src/main/java/io/quarkus/devtools/testing/RegistryClientTest.java:7]: Now anything that uses this class will be setup to use the quarkus-bom platform.

Comment (ia3andy) [reply independent-projects/tools/devtools-testing/src/main/java/io/quarkus/devtools/testing/RegistryClientTest.java:7]: let me try without it to see the result

File: devtools/project-core-extension-codestarts/src/test/java/io/quarkus/devtools/codestarts/quarkus/FunqyHttpCodestartTest.java (hunk 1 above)
Comment (ia3andy) [reply devtools/project-core-extension-codestarts/src/test/java/io/quarkus/devtools/codestarts/quarkus/FunqyHttpCodestartTest.java:23]: Yeah, I use it to make sure the codestart is building, but it's then done in the IT by CI..

File: independent-projects/tools/devtools-testing/src/main/java/io/quarkus/devtools/testing/RegistryClientTest.java (hunk 2 above)
Comment (ia3andy) [reply independent-projects/tools/devtools-testing/src/main/java/io/quarkus/devtools/testing/RegistryClientTest.java:7]: Ok I am not using it anymore from the devtools-testing tests

File: independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/utils/NestedMaps.java [hunk 3]
```diff
@@ -46,10 +46,22 @@ public static void deepMerge(Map left, Map right) {
                 c.addAll((Collection) leftValue);
                 c.addAll((Collection) rightValue);
                 left.put(key, c);
-            } else {
```
Comment (ia3andy) [inline independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/utils/NestedMaps.java:49]: @aloubyansky I found this nasty bug (from myself) while using a static codestart catalog from the tests 😱 the codestart catalog data was muted when generating a project (luckily in prod it's read at each generation).. This is not very optimized, at some point I will probably cache it. So better it be fixed :)

Comment (ia3andy) [reply independent-projects/tools/codestarts/src/main/java/io/quarkus/devtools/codestarts/utils/NestedMaps.java:49]: FYI in this `deepMerge`, it's ok to mutate left, but not right so I added some safety..

**Comment by ia3andy** [comment]
@gsmet each new rebase, I get some new CI errors... I think it's time to do another pass like we did a few months ago, it's been a while the CI haven't been stable.. wdyt?

**Review by aloubyansky — APPROVED** [review]
I have some concerns using `999-SNAPSHOT` in the tests that don't actually rely on the actual Quarkus version being built. It'd be good to change it to some unrelated version in a subsequent PR.

**Pull request merged**